// venom-bot/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.create = exports.connect = exports.Whatsapp = exports.InterfaceState = exports.InterfaceMode = exports.SocketState = exports.MessageType = exports.GroupNotificationType = exports.GroupChangeEvent = exports.ChatState = exports.AckType = void 0;
__exportStar(require("./api/model"), exports);
var enum_1 = require("./api/model/enum");
Object.defineProperty(exports, "AckType", { enumerable: true, get: function () { return enum_1.AckType; } });
Object.defineProperty(exports, "ChatState", { enumerable: true, get: function () { return enum_1.ChatState; } });
Object.defineProperty(exports, "GroupChangeEvent", { enumerable: true, get: function () { return enum_1.GroupChangeEvent; } });
Object.defineProperty(exports, "GroupNotificationType", { enumerable: true, get: function () { return enum_1.GroupNotificationType; } });
Object.defineProperty(exports, "MessageType", { enumerable: true, get: function () { return enum_1.MessageType; } });
Object.defineProperty(exports, "SocketState", { enumerable: true, get: function () { return enum_1.SocketState; } });
Object.defineProperty(exports, "InterfaceMode", { enumerable: true, get: function () { return enum_1.InterfaceMode; } });
Object.defineProperty(exports, "InterfaceState", { enumerable: true, get: function () { return enum_1.InterfaceState; } });
var whatsapp_1 = require("./api/whatsapp");
Object.defineProperty(exports, "Whatsapp", { enumerable: true, get: function () { return whatsapp_1.Whatsapp; } });
var init_1 = require("./controllers/init");
Object.defineProperty(exports, "connect", { enumerable: true, get: function () { return init_1.connect; } });
var initializer_1 = require("./controllers/initializer");
Object.defineProperty(exports, "create", { enumerable: true, get: function () { return initializer_1.create; } });
