// venom-bot/dist/api/helpers/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.dowloadMetaFileBase64 = exports.sleep = exports.loadForceConnect = exports.checkingCloses = exports.callbackWile = exports.deleteFiles = exports.scrapeDeleteToken = exports.scrapeDesconnected = exports.scrapeLogin = exports.scrapeImg = exports.resizeImg = exports.stickerSelect = exports.downloadFileToBase64 = exports.base64MimeType = exports.fileToBase64 = void 0;
var file_to_base64_1 = require("./file-to-base64");
Object.defineProperty(exports, "fileToBase64", { enumerable: true, get: function () { return file_to_base64_1.fileToBase64; } });
var base64_mimetype_1 = require("./base64-mimetype");
Object.defineProperty(exports, "base64MimeType", { enumerable: true, get: function () { return base64_mimetype_1.base64MimeType; } });
var download_file_1 = require("./download-file");
Object.defineProperty(exports, "downloadFileToBase64", { enumerable: true, get: function () { return download_file_1.downloadFileToBase64; } });
var select_sticker_1 = require("./select-sticker");
Object.defineProperty(exports, "stickerSelect", { enumerable: true, get: function () { return select_sticker_1.stickerSelect; } });
Object.defineProperty(exports, "resizeImg", { enumerable: true, get: function () { return select_sticker_1.resizeImg; } });
var scrape_img_qr_1 = require("./scrape-img-qr");
Object.defineProperty(exports, "scrapeImg", { enumerable: true, get: function () { return scrape_img_qr_1.scrapeImg; } });
var scrape_login_1 = require("./scrape-login");
Object.defineProperty(exports, "scrapeLogin", { enumerable: true, get: function () { return scrape_login_1.scrapeLogin; } });
var scrape_desconnect_1 = require("./scrape-desconnect");
Object.defineProperty(exports, "scrapeDesconnected", { enumerable: true, get: function () { return scrape_desconnect_1.scrapeDesconnected; } });
var scrape_deletetoken_1 = require("./scrape-deletetoken");
Object.defineProperty(exports, "scrapeDeleteToken", { enumerable: true, get: function () { return scrape_deletetoken_1.scrapeDeleteToken; } });
var delete_file_1 = require("./delete-file");
Object.defineProperty(exports, "deleteFiles", { enumerable: true, get: function () { return delete_file_1.deleteFiles; } });
var callback_wile_1 = require("./callback-wile");
Object.defineProperty(exports, "callbackWile", { enumerable: true, get: function () { return callback_wile_1.callbackWile; } });
var closes_browser_1 = require("./closes-browser");
Object.defineProperty(exports, "checkingCloses", { enumerable: true, get: function () { return closes_browser_1.checkingCloses; } });
var force_connect_1 = require("./force-connect");
Object.defineProperty(exports, "loadForceConnect", { enumerable: true, get: function () { return force_connect_1.loadForceConnect; } });
var sleep_1 = require("./sleep");
Object.defineProperty(exports, "sleep", { enumerable: true, get: function () { return sleep_1.sleep; } });
var dowload_meta_1 = require("./dowload-meta");
Object.defineProperty(exports, "dowloadMetaFileBase64", { enumerable: true, get: function () { return dowload_meta_1.dowloadMetaFileBase64; } });
